File: apps/android/feature/exercises/domain/src/main/java/com/scribblefit/feature/exercises/domain/usecase/GetExerciseTrendDataUseCase.kt
package com.scribblefit.feature.exercises.domain.usecase

import com.scribblefit.core.common.Calculations
import com.scribblefit.core.common.runCatchingWithCancellation
import com.scribblefit.core.model.Exercise
import com.scribblefit.core.model.TrendDirection
import com.scribblefit.feature.exercises.domain.ExerciseRepository
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.flow.map
import java.util.Calendar

/**
 * Time periods for filtering trend data.
 */
enum class TrendPeriod {
    ONE_MONTH,
    THREE_MONTHS,
    SIX_MONTHS,
    ONE_YEAR,
    ALL
}

/**
 * Data point for a trend chart.
 */
data class TrendDataPoint(
    val date: Long,
    val value: Float
)

/**
 * Insights calculated from trend data.
 */
data class TrendInsights(
    val personalBest: Float,
    val percentageChange: Float,
    val trendDirection: TrendDirection
)

/**
 * Trend data for a specific metric.
 */
data class MetricTrendData(
    val dataPoints: List<TrendDataPoint>,
    val insights: TrendInsights
)

/**
 * Result model containing all data for the trends screen.
 */
data class ExerciseTrendResult(
    val oneRM: MetricTrendData,
    val volume: MetricTrendData
)

/**
 * Use case to fetch and calculate trend data for an exercise.
 */
class GetExerciseTrendDataUseCase(
    private val exerciseRepository: ExerciseRepository,
    private val coroutineDispatcher: CoroutineDispatcher
) {
    operator fun invoke(
        exerciseName: String,
        period: TrendPeriod
    ): Flow<Result<ExerciseTrendResult>> = exerciseRepository.getExercisesByNameFlow(exerciseName)
        .map { allHistory ->
            runCatchingWithCancellation {
                val sortedHistory = allHistory.sortedBy { it.createdAt }

                if (sortedHistory.isEmpty()) {
                    val emptyData = MetricTrendData(
                        emptyList(),
                        TrendInsights(0f, 0f, TrendDirection.STABLE)
                    )
                    return@runCatchingWithCancellation ExerciseTrendResult(
                        oneRM = emptyData,
                        volume = emptyData
                    )
                }

                val filteredHistory = filterByPeriod(sortedHistory, period)

                val oneRMData = calculateMetricData(filteredHistory) { exercise ->
                    exercise.sets.maxOfOrNull {
                        Calculations.calculate1RM(it.weight ?: 0f, it.reps)
                    } ?: 0f
                }

                val volumeData = calculateMetricData(filteredHistory) { exercise ->
                    exercise.sets.sumOf {
                        Calculations.calculateVolume(it.weight, it.reps).toDouble()
                    }.toFloat()
                }

                ExerciseTrendResult(
                    oneRM = oneRMData,
                    volume = volumeData
                )
            }
        }
        .flowOn(coroutineDispatcher)

    private fun calculateMetricData(
        history: List<Exercise>,
        valueExtractor: (Exercise) -> Float
    ): MetricTrendData {
        val dataPoints = history.map { exercise ->
            TrendDataPoint(exercise.createdAt, valueExtractor(exercise))
        }

        val personalBest = dataPoints.maxOfOrNull { it.value } ?: 0f

        // Calculate percentage change between start and end of period
        val percentageChange = if (dataPoints.size >= 2) {
            val first = dataPoints.first().value
            val last = dataPoints.last().value
            if (first > 0) ((last - first) / first) * 100f else 0f
        } else 0f

        val direction = when {
            percentageChange > 5f -> TrendDirection.IMPROVING
            percentageChange < -5f -> TrendDirection.DECLINING
            else -> TrendDirection.STABLE
        }

        return MetricTrendData(
            dataPoints = dataPoints,
            insights = TrendInsights(personalBest, percentageChange, direction)
        )
    }

    private fun filterByPeriod(
        history: List<Exercise>,
        period: TrendPeriod
    ): List<Exercise> {
        if (period == TrendPeriod.ALL) return history

        val calendar = Calendar.getInstance()
        when (period) {
            TrendPeriod.ONE_MONTH -> calendar.add(Calendar.MONTH, -1)
            TrendPeriod.THREE_MONTHS -> calendar.add(Calendar.MONTH, -3)
            TrendPeriod.SIX_MONTHS -> calendar.add(Calendar.MONTH, -6)
            TrendPeriod.ONE_YEAR -> calendar.add(Calendar.YEAR, -1)
        }
        val startTime = calendar.timeInMillis
        return history.filter { it.createdAt >= startTime }
    }
}

File: apps/ios/ScribbleFit/LocalPackages/Features/FeatureExercises/Sources/FeatureExercises/Domain/UseCases/GetExerciseTrendDataUseCase.swift
import Foundation
import CoreModel
import CoreCommon

/**
 * Time periods for filtering trend data.
 */
public enum TrendPeriod: String, Sendable, Codable {
    case oneMonth = "1M"
    case threeMonths = "3M"
    case sixMonths = "6M"
    case oneYear = "1Y"
    case all = "ALL"
}

/**
 * Data point for a trend chart.
 */
public struct TrendDataPoint: Sendable, Identifiable {
    public let id = UUID()
    public let date: Date
    public let value: Float
}

/**
 * Insights calculated from trend data.
 */
public struct TrendInsights: Sendable {
    public let personalBest: Float
    public let percentageChange: Float
    public let trendDirection: TrendDirection
}

/**
 * Trend data for a specific metric.
 */
public struct MetricTrendData: Sendable {
    public let dataPoints: [TrendDataPoint]
    public let insights: TrendInsights
}

/**
 * Result model containing all data for the trends screen.
 */
public struct ExerciseTrendResult: Sendable {
    public let oneRM: MetricTrendData
    public let volume: MetricTrendData
}

/**
 * Use case to fetch and calculate trend data for an exercise.
 */
@MainActor
public final class GetExerciseTrendDataUseCase {
    private let exerciseRepository: ExerciseRepository
    
    public init(exerciseRepository: ExerciseRepository) {
        self.exerciseRepository = exerciseRepository
    }
    
    public func execute(
        exerciseName: String,
        period: TrendPeriod
    ) -> AsyncStream<ExerciseTrendResult> {
        AsyncStream { continuation in
            let exercisesStream = exerciseRepository.observeExercises(query: exerciseName)
            
            let task = Task {
                for await exercises in exercisesStream {
                    let allHistory = exercises
                        .filter { $0.canonicalName.lowercased() == exerciseName.lowercased() }
                        .sorted(by: { $0.createdAt < $1.createdAt })
                    
                    if allHistory.isEmpty {
                        let emptyData = MetricTrendData(
                            dataPoints: [],
                            insights: TrendInsights(personalBest: 0, percentageChange: 0, trendDirection: .stable)
                        )
                        continuation.yield(ExerciseTrendResult(
                            oneRM: emptyData,
                            volume: emptyData
                        ))
                        continue
                    }
                    
                    let filteredHistory = filterByPeriod(allHistory, period: period)
                    
                    let oneRMData = calculateMetricData(filteredHistory) { exercise in
                        exercise.sets.map { Calculations.calculate1RM(weight: $0.weight ?? 0, reps: $0.reps) }.max() ?? 0
                    }
                    
                    let volumeData = calculateMetricData(filteredHistory) { exercise in
                        exercise.sets.reduce(Float(0.0)) { $0 + Calculations.calculateVolume(weight: $1.weight, reps: $1.reps) }
                    }
                    
                    continuation.yield(ExerciseTrendResult(
                        oneRM: oneRMData,
                        volume: volumeData
                    ))
                }
            }
            
            continuation.onTermination = { _ in
                task.cancel()
            }
        }
    }
    
    private func calculateMetricData(
        _ history: [Exercise],
        valueExtractor: (Exercise) -> Float
    ) -> MetricTrendData {
        let dataPoints = history.map { exercise in
            TrendDataPoint(date: exercise.createdAt, value: valueExtractor(exercise))
        }
        
        let personalBest = dataPoints.map { $0.value }.max() ?? 0
        
        let percentageChange: Float = {
            guard dataPoints.count >= 2, let first = dataPoints.first?.value, first > 0 else { return 0 }
            let last = dataPoints.last?.value ?? 0
            return ((last - first) / first) * 100
        }()
        
        let direction: TrendDirection = {
            if percentageChange > 5 { return .improving }
            if percentageChange < -5 { return .declining }
            return .stable
        }()
        
        return MetricTrendData(
            dataPoints: dataPoints,
            insights: TrendInsights(personalBest: personalBest, percentageChange: percentageChange, trendDirection: direction)
        )
    }
    
    private func filterByPeriod(_ history: [Exercise], period: TrendPeriod) -> [Exercise] {
        if period == .all { return history }
        
        let calendar = Calendar.current
        let now = Date()
        let startDate: Date? = {
            switch period {
            case .oneMonth: return calendar.date(byAdding: .month, value: -1, to: now)
            case .threeMonths: return calendar.date(byAdding: .month, value: -3, to: now)
            case .sixMonths: return calendar.date(byAdding: .month, value: -6, to: now)
            case .oneYear: return calendar.date(byAdding: .year, value: -1, to: now)
            case .all: return nil
            }
        }()
        
        guard let start = startDate else { return history }
        return history.filter { $0.createdAt >= start }
    }
}
